identity-account-storage/src/storage/tests.rs

```diff
@@ -0,0 +1,503 @@
+  let network: NetworkName = Network::Mainnet.name();
+
+  let (did, _): (IotaDID, _) = storage
+    .did_create(network.clone(), &fragment, None)
+    .await
+    .context("did_create returned an error")?;
+
+  let mut locations: Vec<KeyLocation> = Vec::with_capacity(NUM_IDENTITIES);
+
+  for _ in 0..NUM_IDENTITIES {
+    let key_fragment: String = random_string();
+    let location: KeyLocation = storage
+      .key_generate(&did, KeyType::Ed25519, &key_fragment)
+      .await
+      .context("key_generate returned an error")?;
+    locations.push(location);
+  }
+
+  for location in locations {
+    let exists: bool = storage
+      .key_exists(&did, &location)
+      .await
+      .context("key_exists returned an error")?;
+
+    ensure!(exists, "expected key at location `{location}` to exist");
+
+    let deleted: bool = storage
+      .key_delete(&did, &location)
+      .await
+      .context("key_delete returned an error")?;
+
+    ensure!(deleted, "expected key at location `{location}` to be deleted");
+
+    let deleted: bool = storage
+      .key_delete(&did, &location)
+      .await
+      .context("key_delete returned an error")?;
+
+    ensure!(!deleted, "expected key at location `{location}` to already be deleted");
+  }
+
+  Ok(())
+}
+
+#[named]
+pub async fn storage_did_list_test(storage: Box<dyn Storage>) -> anyhow::Result<()> {
+  const NUM_IDENTITIES: usize = 20;
+  let fragment: String = random_string();
+  let network: NetworkName = Network::Mainnet.name();
+
+  let list: Vec<IotaDID> = storage.did_list().await.context("did_list returned an error")?;
+
+  ensure!(
+    list.is_empty(),
+    "expected list to be empty, but found {} element(s)",
+    list.len()
+  );
+
+  let mut dids: Vec<IotaDID> = Vec::with_capacity(NUM_IDENTITIES);
+  for i in 0..NUM_IDENTITIES {
+    let (did, _): (IotaDID, _) = storage
+      .did_create(network.clone(), &fragment, None)
+      .await
+      .context("did_create returned an error")?;
+
+    let exists: bool = storage.did_exists(&did).await.context("did_exists returned an error")?;
+    ensure!(exists, "expected did `{did}` to exist");
+
+    let list_len: usize = storage.did_list().await.context("did_list returned an error")?.len();
+    let expected_len: usize = i + 1;
+
+    ensure_eq!(
+      list_len,
+      expected_len,
+      "expected did_list to return a list of len {expected_len}, got {list_len} elements instead"
+    );
+
+    dids.push(did);
+  }
+
+  for (i, did) in dids.into_iter().enumerate() {
+    let purged: bool = storage.did_purge(&did).await.context("did_purge returned an error")?;
+
+    ensure!(purged, "expected the did `{did}` to be purged");
+
+    let exists: bool = storage.did_exists(&did).await.context("did_exists returned an error")?;
+    ensure!(!exists, "expected did `{did}` to no longer exist");
+
+    let list_len: usize = storage.did_list().await.context("did_list returned an error")?.len();
+    let expected_len: usize = NUM_IDENTITIES - (i + 1);
```

Comment:
This is kind of prevalent throughout the tests but calling other interface function makes it slightly more difficult for implementers to test functions independently (but I consider that mostly fine since these are meant to be comprehensive). I'm only mentioning it here because in this case `did_purge` should not be required for the `did_list` function, since `Storage` implementers may choose not to implement `did_purge` at all (for various reasons).

Suggestion: move these checks to the `did_purge` test rather.